File: core/src/com/jediterm/terminal/TextStyle.java
package com.jediterm.terminal;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.ref.WeakReference;
import java.util.EnumSet;
import java.util.Objects;
import java.util.WeakHashMap;

public class TextStyle {
  private static final EnumSet<Option> NO_OPTIONS = EnumSet.noneOf(Option.class);

  public static final TextStyle EMPTY = new TextStyle();

  private static final WeakHashMap<TextStyle, WeakReference<TextStyle>> styles = new WeakHashMap<>();

  private final TerminalColor myForeground;
  private final TerminalColor myBackground;
  private final EnumSet<Option> myOptions;

  public TextStyle() {
    this(null, null, NO_OPTIONS);
  }

  public TextStyle(@Nullable TerminalColor foreground, @Nullable TerminalColor background) {
    this(foreground, background, NO_OPTIONS);
  }

  public TextStyle(@Nullable TerminalColor foreground, @Nullable TerminalColor background, @NotNull EnumSet<Option> options) {
    myForeground = foreground;
    myBackground = background;
    myOptions = options.clone();
  }

  @NotNull
  public static TextStyle getCanonicalStyle(TextStyle currentStyle) {
    if (currentStyle instanceof HyperlinkStyle) {
      return currentStyle;
    }
    final WeakReference<TextStyle> canonRef = styles.get(currentStyle);
    if (canonRef != null) {
      final TextStyle canonStyle = canonRef.get();
      if (canonStyle != null) {
        return canonStyle;
      }
    }
    styles.put(currentStyle, new WeakReference<>(currentStyle));
    return currentStyle;
  }

  @Nullable
  public TerminalColor getForeground() {
    return myForeground;
  }

  @Nullable
  public TerminalColor getBackground() {
    return myBackground;
  }

  public TextStyle createEmptyWithColors() {
    return new TextStyle(myForeground, myBackground);
  }

  public int getId() {
    return hashCode();
  }

  public boolean hasOption(final Option option) {
    return myOptions.contains(option);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    TextStyle textStyle = (TextStyle) o;
    return Objects.equals(myForeground, textStyle.myForeground) &&
      Objects.equals(myBackground, textStyle.myBackground) &&
      myOptions.equals(textStyle.myOptions);
  }

  @Override
  public int hashCode() {
    return Objects.hash(myForeground, myBackground, myOptions);
  }

  public TerminalColor getBackgroundForRun() {
    return myOptions.contains(Option.INVERSE) ? myForeground : myBackground;
  }

  public TerminalColor getForegroundForRun() {
    return myOptions.contains(Option.INVERSE) ? myBackground : myForeground;
  }

  @NotNull
  public Builder toBuilder() {
    return new Builder(this);
  }

  public enum Option {
    BOLD,
    ITALIC,
    BLINK,
    DIM,
    INVERSE,
    UNDERLINED,
    HIDDEN;

    private void set(@NotNull EnumSet<Option> options, boolean val) {
      if (val) {
        options.add(this);
      }
      else {
        options.remove(this);
      }
    }
  }

  public static class Builder {
    private TerminalColor myForeground;
    private TerminalColor myBackground;
    private EnumSet<Option> myOptions;

    public Builder(@NotNull TextStyle textStyle) {
      myForeground = textStyle.myForeground;
      myBackground = textStyle.myBackground;
      myOptions = textStyle.myOptions.clone();
    }

    public Builder() {
      myForeground = null;
      myBackground = null;
      myOptions = EnumSet.noneOf(Option.class);
    }

    @NotNull
    public Builder setForeground(@Nullable TerminalColor foreground) {
      myForeground = foreground;
      return this;
    }

    @NotNull
    public Builder setBackground(@Nullable TerminalColor background) {
      myBackground = background;
      return this;
    }

    @NotNull
    public Builder setOption(@NotNull Option option, boolean val) {
      option.set(myOptions, val);
      return this;
    }

    @NotNull
    public TextStyle build() {
      return new TextStyle(myForeground, myBackground, myOptions);
    }
  }
}

File: bossterm-core-mpp/src/jvmMain/kotlin/ai/rever/bossterm/terminal/TextStyle.kt
package ai.rever.bossterm.terminal

import java.util.*

open class TextStyle @JvmOverloads constructor(
    val foreground: TerminalColor? = null,
    val background: TerminalColor? = null,
    options: EnumSet<Option> = NO_OPTIONS
) {
    private val myOptions: EnumSet<Option>

    init {
        myOptions = options.clone()
    }

    fun createEmptyWithColors(): TextStyle {
        return TextStyle(this.foreground, this.background)
    }

    fun hasOption(option: Option?): Boolean {
        return myOptions.contains(option)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val textStyle = other as TextStyle
        return this.foreground == textStyle.foreground &&
                this.background == textStyle.background &&
                myOptions == textStyle.myOptions
    }

    override fun hashCode(): Int {
        return Objects.hash(this.foreground, this.background, myOptions)
    }

    open fun toBuilder(): Builder {
        return Builder(this)
    }

    enum class Option {
        BOLD,
        ITALIC,
        SLOW_BLINK,
        RAPID_BLINK,
        DIM,
        INVERSE,
        UNDERLINED,
        HIDDEN,
        PROTECTED;

        fun set(options: EnumSet<Option>, `val`: Boolean) {
            if (`val`) {
                options.add(this)
            } else {
                options.remove(this)
            }
        }
    }

    open class Builder {
        private var myForeground: TerminalColor?
        private var myBackground: TerminalColor?
        private val myOptions: EnumSet<Option>

        constructor(textStyle: TextStyle) {
            myForeground = textStyle.foreground
            myBackground = textStyle.background
            myOptions = textStyle.myOptions.clone()
        }

        constructor() {
            myForeground = null
            myBackground = null
            myOptions = EnumSet.noneOf(Option::class.java)
        }

        fun setForeground(foreground: TerminalColor?): Builder {
            myForeground = foreground
            return this
        }

        fun setBackground(background: TerminalColor?): Builder {
            myBackground = background
            return this
        }

        fun setOption(option: Option, `val`: Boolean): Builder {
            option.set(myOptions, `val`)
            return this
        }

        open fun build(): TextStyle {
            return TextStyle(myForeground, myBackground, myOptions)
        }
    }

    companion object {
        private val NO_OPTIONS: EnumSet<Option> = EnumSet.noneOf(Option::class.java)

        val EMPTY: TextStyle = TextStyle()
    }
}
